src/core/public/ui_settings/ui_settings_client.ts

```diff
@@ -0,0 +1,251 @@
+   * setting, and error behavior of that method.
+   */
+  public async remove(key: string) {
+    return await this.update(key, null);
+  }
+
+  /**
+   * Returns true if the key is a "known" uiSetting, meaning it is either defined in the
+   * uiSettingDefaults or was previously added as a custom setting via the `set()` method.
+   */
+  public isDeclared(key: string) {
+    return Boolean(key in this.cache);
+  }
+
+  /**
+   * Returns true if the setting has no user-defined value or is unknown
+   */
+  public isDefault(key: string) {
+    return !this.isDeclared(key) || this.cache[key].userValue == null;
+  }
+
+  /**
+   * Returns true if the setting is not a part of the uiSettingDefaults, but was either
+   * added directly via `set()`, or is an unknown setting found in the uiSettings saved
+   * object
+   */
+  public isCustom(key: string) {
+    return this.isDeclared(key) && !('value' in this.cache[key]);
+  }
+
+  /**
+   * Returns true if a settings value is overridden by the server. When a setting is overridden
+   * its value can not be changed via `set()` or `remove()`.
+   */
+  public isOverridden(key: string) {
+    return this.isDeclared(key) && Boolean(this.cache[key].isOverridden);
+  }
+
+  /**
+   * Overrides the default value for a setting in this specific browser tab. If the page
+   * is reloaded the default override is lost.
+   */
+  public overrideLocalDefault(key: string, newDefault: any) {
+    // capture the previous value
+    const prevDefault = this.defaults[key] ? this.defaults[key].value : undefined;
+
+    // update defaults map
+    this.defaults[key] = {
+      ...(this.defaults[key] || {}),
+      value: newDefault,
+    };
+
+    // update cached default value
+    this.cache[key] = {
+      ...(this.cache[key] || {}),
+      value: newDefault,
+    };
+
+    // don't broadcast change if userValue was already overriding the default
+    if (this.cache[key].userValue == null) {
+      this.update$.next({
+        key,
+        newValue: newDefault,
+        oldValue: prevDefault,
+      });
+    }
+  }
+
+  /**
+   * Returns an Observable that notifies subscribers of each update to the uiSettings,
+   * including the key, newValue, and oldValue of the setting that changed.
+   */
+  public getUpdate$() {
+    return this.update$.asObservable();
+  }
+
+  /**
+   * Prepares the uiSettingsClient to be discarded, completing any update$ observerables
+   * that have been created.
+   */
+  public stop() {
+    this.update$.complete();
+  }
+
+  private assertUpdateAllowed(key: string) {
+    if (this.isOverridden(key)) {
+      throw new Error(
+        `Unable to update "${key}" because its value is overridden by the Kibana server`
+      );
+    }
+  }
+
+  private async update(key: string, newVal: any) {
+    this.assertUpdateAllowed(key);
+
+    const declared = this.isDeclared(key);
+    const defaults = this.defaults;
+
+    const oldVal = declared ? this.cache[key].userValue : undefined;
```

Comment:
We didn't, I checked the code since it was first introduced and best I can tell `defaults[key].defaultValue` was never actually defined, so I just got rid of it.